File: Subarray Product Less Than K.cpp
class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
        if (k == 0) return 0;
        int n = nums.size();
        double k2 = log((double) k);
        vector<double> arr(n, 0);
        for (int i = 0; i < n; i++) {
            arr[i] = log((double) nums[i]);
        }

        vector<double> dp(n+1, 0);
        double summ = 0;
        int count = 0;
        for (int i = 0; i < n; i++) {
            summ += arr[i];
            dp[i+1] = summ;
        }

        for (int i = 0; i < n; i++) {
            double remainder = k2 + dp[i];
            auto j = lower_bound(dp.begin(), dp.end(), remainder);
            int j2 = j - dp.begin();
            if (j2 < i+1) continue;
            count += j2-i-1;
        }

        return count;
    }
};

File: Subarray Product Less Than K.py
class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        if k == 0:
            return 0
        n = len(nums)
        k = math.log(k)
        nums = [math.log(num) for num in nums]
        dp = [0] * (n + 1)
        summ = 0
        count = 0
        for i, num in enumerate(nums):
            summ += num
            dp[i + 1] = summ

        for i, num in enumerate(nums):
            remainder = k + dp[i]
            j = bisect.bisect_left(dp, remainder)
            if j < i + 1:
                continue

            count += j - i - 1

        return count
